### PR TITLE
Core/poc retry transaction2

### PR DESCRIPTION
**Pull request by stevencl840**
[sc-30139]

### CODE REVIEW

File: source/Nevermore/Util/DataModificationQueryBuilder.cs [hunk 1]
```diff
@@ -342,7 +342,7 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, object> allocateI
             switch (mapping.JsonStorageFormat)
             {
                 case JsonStorageFormat.TextOnly:
-                    result[$"{prefix}{JsonVariableName}"] = serializer.SerializeText(document, mapping);
+                    result[$"{prefix}{JsonVariableName}"] = serializer.SerializeText(document, mapping).ReadToEnd();
```
Comment (stevencl840) [inline source/Nevermore/Util/DataModificationQueryBuilder.cs:345]: Can't rewind textreaders, so as a work around for the POC I just output the string. Will need to come up with something better if we go ahead with this

File: source/Nevermore/CommandExecutor.cs [hunk 2]
```diff
@@ -258,7 +258,7 @@ void DisposeOfParameters()
                 // Parameters can contain streams and other disposable objects.
                 if (parameter.Value is IDisposable disposable)
                 {
-                    disposable.Dispose();
+                    //disposable.Dispose();
```
Comment (stevencl840) [inline source/Nevermore/CommandExecutor.cs:261]: We can dispose of these at the transaction level once we are done. Just left it like this as it is only a PoC

File: source/Nevermore/Transient/SqlDatabaseTransientErrorDetectionStrategy.cs [hunk 3]
```diff
@@ -220,6 +220,7 @@ public bool IsTransient(Exception ex)
             }
 
             // Otherwise it could be another simple transient error
+            var result = sqlErrors.Select(x => x.Number).Intersect(SimpleTransientErrorCodes).Any();
```
Comment (stevencl840) [inline source/Nevermore/Transient/SqlDatabaseTransientErrorDetectionStrategy.cs:223]: Clean me

File: source/Nevermore/Advanced/ReadTransaction.cs [hunk 4]
```diff
@@ -714,7 +767,9 @@ RetryPolicy GetRetryPolicy(RetriableOperation operation)
             if (operationsToRetry == RetriableOperation.None)
                 return RetryPolicy.NoRetry;
 
-            return (operationsToRetry & operation) != 0 ? RetryManager.Instance.GetDefaultSqlCommandRetryPolicy() : RetryPolicy.NoRetry;
+            //return (operationsToRetry & operation) != 0 ? RetryManager.Instance.GetDefaultSqlCommandRetryPolicy() : RetryPolicy.NoRetry;
+
+            return RetryManager.Instance.GetDefaultSqlCommandRetryPolicy();
```
Comment (stevencl840) [inline source/Nevermore/Advanced/ReadTransaction.cs:772]: We are not rolling back on an operation, instead a whole transaction so I short circuited this  for the purposed of POC

File: source/Nevermore/CommandExecutor.cs [hunk 5]
```diff
@@ -42,13 +42,13 @@ public CommandExecutor(DbCommand command, PreparedCommand prepared, RetryPolicy
...
 
-        public int ExecuteNonQuery()
+        public int ExecuteNonQuery(Action replayClosure)
         {
             AssertSynchronousOperation();
             try
             {
                 AssertSynchronousOperation();
-                return command.ExecuteNonQueryWithRetry(retryPolicy);
+                return command.ExecuteNonQueryWithRetry(retryPolicy, replayClosure);
```
Comment (akirayamamoto) [inline source/Nevermore/CommandExecutor.cs:51]: Interesting approach :)

**Review by akirayamamoto — COMMENTED** [review]
Nice approach, it looks like this PoC will make use of the `GetRetryPolicy` method which means we will re-use the existing policy code as we discussed.

File: source/Nevermore/Advanced/ReadTransaction.cs [hunk 6]
```diff
@@ -565,23 +644,47 @@ public int ExecuteNonQuery(string query, CommandParameterValues? args = null, Ti
...
 
         public int ExecuteNonQuery(PreparedCommand preparedCommand)
         {
             using var mutex = DeadlockAwareLock.Lock();
-            using var command = CreateCommand(preparedCommand);
-            return command.ExecuteNonQuery();
+            var retryPolicy = GetRetryPolicy(preparedCommand.Operation).LoggingRetries(nameof(ExecuteNonQuery));
+            var result = retryPolicy.ExecuteAction(() =>
+            {
+                if (connection?.State is ConnectionState.Closed)
```
Comment (stevencl840) [inline source/Nevermore/Advanced/ReadTransaction.cs:658]: Need to investigate if this is sufficient

Comment (akirayamamoto) [reply source/Nevermore/Advanced/ReadTransaction.cs:658]: I wonder if we can make use of `@@TRANCOUNT == 1`. Will we ever have multiple transactions open one on top of the other?

**Pull request closed without merging**